### PR TITLE
Fix IMAs in Flash-Attention splitkv kernel

### PR DESCRIPTION
**Pull request by drisspg**
# Summary

While debugging CI failures for flash_attention tests I stumbled across 2 IMAs for the split-kv variant of flash attention.
1. Illegal global memory writes during the writing of softmax_lse_accum. This was pinpointed to the temporary liftime of these out_accum and softmax_lse_accum. These were likely getting their refcount dropped **before** the kernel launch that used, them allowing them to potentially get used for other allocations.
2. After debugging this there was illegal writes of the combine kernel. I was able to pinpoint this to the writing to the reduce LSE. From my understanding it was making assumption that kBlocKM evenly divided the global number of rows and wasn't masking out these writes.

### History
My line of thinking for this: 

We create the temporary split accum + LSE stats tensors to store the data for each split. We then launch a follow up kernel to do the reduction.

Under ordinary non roofline memory usage the cuda memory caching allocator will keep these allocations alive even though the tensors were created within a temporary scope and no longer have any live references.

On CI we often run near max memory usage. We change/add tests and suddenly we get close to oom threshold. The memory allocator will reap these segments and we get write after free errors.

After that  fix I did get further past the splitkv_flash kernel and then got the following error:

``` Shell
❯ TORCH_DISABLE_ADDR2LINE=1 PYTORCH_NO_CUDA_MEMORY_CACHING=1  compute-sanitizer --show-backtrace=device --tool memcheck --log-file ima.txt python ima.py

softmax_lseaccum_ptr =0x7f5ebb208a00
oaccum_ptr =0x7f5ebb208c00
softmax_lse_ptr = 0x7f5ebb208800
❯ 
❯ head ima.txt -n 10
========= COMPUTE-SANITIZER
========= Invalid __global__ write of size 4 bytes
=========     at void pytorch_flash::flash_fwd_splitkv_combine_kernel<pytorch_flash::Flash_fwd_kernel_traits<(int)32, (int)64, (int)256, (int)4, (bool)0, (bool)0, cutlass::bfloat16_t, pytorch_flash::Flash_kernel_traits<(int)32, (int)64, (int)256, (int)4, cutlass::bfloat16_t>>, (int)16, (int)1, (bool)1>(pytorch_flash::Flash_fwd_params)+0x630
=========     by thread (2,0,0) in block (0,0,0)
=========     Address 0x7f5ebb208804 is out of bounds
=========     and is 1 bytes after the nearest allocation at 0x7f5ebb208800 of size 4 bytes
```

Okay I looked at the address and it looks like we are writing consective bytes past the softmax_lse_ptr in from the combine func: I tried padding out the softmax_lse to q_padded and no more illegal memory errors on my repro:
```
========= COMPUTE-SANITIZER
========= ERROR SUMMARY: 0 errors
```

Fixes https://github.com/pytorch/pytorch/issues/131240
Fixes https://github.com/pytorch/pytorch/issues/131227
Fixes https://github.com/pytorch/pytorch/issues/131221

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131277

### CODE REVIEW

File: aten/src/ATen/native/transformers/cuda/flash_attn/flash_api.cpp [hunk 1]
```diff
@@ -308,18 +307,23 @@ void set_params_splitkv(Flash_fwd_params &params, const int batch_size,
...
         }
         if (params.num_splits > 1) {
-            at::Tensor softmax_lse_accum = at::empty({params.num_splits, batch_size, num_heads, max_seqlen_q}, opts.dtype(at::kFloat));
-            at::Tensor out_accum = at::empty({params.num_splits, batch_size, num_heads, max_seqlen_q, head_size_rounded}, opts.dtype(at::kFloat));
-            params.softmax_lseaccum_ptr = softmax_lse_accum.data_ptr();
-            params.oaccum_ptr = out_accum.data_ptr();
+            softmax_lse_accum = at::empty({params.num_splits, batch_size, num_heads, max_seqlen_q}, opts.dtype(at::kFloat));
+            out_accum = at::empty({params.num_splits, batch_size, num_heads, max_seqlen_q, head_size_rounded}, opts.dtype(at::kFloat));
+            // params.softmax_lseaccum_ptr = softmax_lse_accum.data_ptr();
+            // params.oaccum_ptr = out_accum.data_ptr();
```
Comment (malfet) [inline aten/src/ATen/native/transformers/cuda/flash_attn/flash_api.cpp:321]: This is valid
```suggestion
            params.softmax_lseaccum_ptr = softmax_lse_accum.data_ptr();
            params.oaccum_ptr = out_accum.data_ptr();
```

File: aten/src/ATen/native/transformers/cuda/flash_attn/flash_api.cpp [hunk 2]
```diff
@@ -470,9 +474,28 @@ mha_fwd(const at::Tensor &q,         // batch_size x seqlen_q x num_heads x head
...
+    at::Tensor softmax_lse_accum, out_accum;
+    std::tie(softmax_lse_accum, out_accum) = set_params_splitkv(params, batch_size, num_heads,
+                        head_size, seqlen_k, seqlen_q,
+                        head_size_rounded, p_dropout, /*num_splits*/0, dprops, opts);
+    if (softmax_lse_accum.defined()){
+       params.softmax_lseaccum_ptr = softmax_lse_accum.data_ptr();
+    }
+    if (out_accum.defined()){
+       params.oaccum_ptr = out_accum.data_ptr();
+    }
```
Comment (malfet) [inline aten/src/ATen/native/transformers/cuda/flash_attn/flash_api.cpp:487]: Those are superfluous, can be safely done in set_params_splitkv

**Review by malfet — APPROVED** [review]
This move in the right direction, but I don't think it fully solves the problem, because afaik there are no guarantee that softmax_lse_accum will not be reused when fwd kernel is actually executed on the stream.

I wonder if compute_sanitizer inserts an explicit syncs after every kernel call

**Comment by drisspg** [comment]
@ezyang on the failing test I am getting this tlparse: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/.tmpZCnkrl/index.html

I am not sure what the best way forward is

**Comment by drisspg** [comment]
Better fix for the combine kernel is to actually not write out here: https://github.com/pytorch/pytorch/blob/f8f41dcb24cb4f4e87a51bb04847942dd835e496/aten/src/ATen/native/transformers/cuda/flash_attn/flash_fwd_kernel.h#L1154

**Comment by drisspg** [comment]
@tridao Would love your eyes on the reduction kernel changes, as you can tell by the naming I still working through the masking for the write.

If looks good I can open a PR to FA repo

**Comment by tridao** [comment]
Oh right, we need masking on the LSE write.
Thanks for the fix and the thorough investigation!
Can you open a PR?

**Comment by drisspg** [comment]
@sgrigory appears to have solved a similar problem in pull 970

I will land this as is and then do a follow up PR to PyTorch updating our fork to include some more recent changes from Flash.
